src/Func.h

```diff
@@ -1351,10 +1351,14 @@ class Func {
     }
     // @}
 
-    /** Schedule for execution using GLSL. Conceptually, this is similar to
-     * parallelization over 'x' and 'y' (since GLSL shaders compute individual
-     * output pixels in parallel) and vectorization over 'c' (since GLSL
-     * implicitly vectorizes the color channel). */
+    /** Schedule for execution using coordinate-based hardware api.
+     * GLSL and Renderscript are examples of those. Conceptually, this is
+     * similar to parallelization over 'x' and 'y' (since GLSL shaders compute
+     * individual output pixels in parallel) and vectorization over 'c'
+     * (since GLSL/RS implicitly vectorizes the color channel). */
+    EXPORT Func &image(Var x, Var y, Var c, DeviceAPI device_api);
```

Comment:
I like that "image" avoids putting a backend detail in the schedule API
compared to gpu_threads, etc.  But "image" is an overloaded term and it
takes on another meaning here.

Is it possible to indicate that the directive is constraining the schedule
based on the DeviceAPI, without making an analogy?  You can schedule for
OpenCL with gpu_threads (but OpenCL doesn't have threads- work items are
analogous to threads). Most of the DeviceAPIs have kernels: render script,
cuda, opencl; glsl has shaders.

What about something like: parallel_with(DeviceAPI, Var, Var, ... )

On Tue, Apr 28, 2015 at 11:38 AM, Andrew Adams <notifications@github.com>
wrote:

> In src/Func.h
> <https://github.com/halide/Halide/pull/761#discussion_r29274172>:
>
> > @@ -1351,10 +1351,14 @@ class Func {
> >      }
> >      // @}
> >
> > -    /** Schedule for execution using GLSL. Conceptually, this is similar to
> > -     * parallelization over 'x' and 'y' (since GLSL shaders compute individual
> > -     * output pixels in parallel) and vectorization over 'c' (since GLSL
> > -     * implicitly vectorizes the color channel). */
> > +    /** Schedule for execution using coordinate-based hardware api.
> > +     * GLSL and Renderscript are examples of those. Conceptually, this is
> > +     * similar to parallelization over 'x' and 'y' (since GLSL shaders compute
> > +     * individual output pixels in parallel) and vectorization over 'c'
> > +     * (since GLSL/RS implicitly vectorizes the color channel). */
> > +    EXPORT Func &image(Var x, Var y, Var c, DeviceAPI device_api);
>
> I think "shader" is better than "image".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/761/files#r29274172>.
>